core/include/vecmem/containers/static_vector.hpp

```diff
@@ -7,170 +7,271 @@
+   ///
    template< typename TYPE, std::size_t MAX_SIZE >
    class static_vector {
 
    public:
-      /// @name Type definitions mimicking @c std::vector
+      /// @name Type definitions, mimicking @c std::vector
       /// @{
-      typedef std::size_t    size_type;
-      typedef std::ptrdiff_t difference_type;
-      typedef TYPE*          pointer;
-      typedef const TYPE*    const_pointer;
-      typedef TYPE&          reference;
-      typedef const TYPE&    const_reference;
-      typedef TYPE           value_type;
-      static const size_type array_max_size = MAX_SIZE;
-      typedef TYPE           array_type[ array_max_size ];
-      typedef typename std::vector< TYPE >::iterator         iterator;
-      typedef typename std::vector< TYPE >::const_iterator   const_iterator;
-      typedef typename std::vector< TYPE >::reverse_iterator reverse_iterator;
-      typedef typename std::vector< TYPE >::const_reverse_iterator
-         const_reverse_iterator;
+
+      /// Type of the array elements
+      typedef TYPE               value_type;
+      /// Size type for the array
+      typedef std::size_t        size_type;
+      /// Pointer difference type
+      typedef std::ptrdiff_t     difference_type;
+
+      /// The maximal size of the vector
+      static constexpr size_type array_max_size = MAX_SIZE;
+      /// The size of the vector elements
+      static constexpr size_type value_size = sizeof( value_type );
+      /// Type of the array holding the payload of the vector elements
+      typedef char               array_type[ array_max_size * value_size ];
+
+      /// Value reference type
+      typedef value_type&        reference;
+      /// Constant value reference type
+      typedef const value_type&  const_reference;
+      /// Value pointer type
+      typedef value_type*        pointer;
+      /// Constant value pointer type
+      typedef const value_type*  const_pointer;
+
+      /// Forward iterator type
+      typedef pointer            iterator;
+      /// Constant forward iterator type
+      typedef const_pointer      const_iterator;
+      /// Reverse iterator type
+      typedef vecmem::reverse_iterator< iterator >       reverse_iterator;
+      /// Constant reverse iterator type
+      typedef vecmem::reverse_iterator< const_iterator > const_reverse_iterator;
+
       /// @}
 
+      /// @name Constructors and destructor, mimicking @c std::vector
+      /// @{
+
       /// Default constructor
       VECMEM_HOST_AND_DEVICE
-      static_vector( std::size_t size = 0 ) : m_size( size ) {}
+      static_vector();
+      /// Construct a vector with a specific size
+      VECMEM_HOST_AND_DEVICE
+      static_vector( size_type size, const_reference value = value_type() );
+      /// Construct a vector with values coming from a pair of iterators
+      /*
+      template< typename InputIt >
+      VECMEM_HOST_AND_DEVICE
+      static_vector( InputIt other_begin, InputIt other_end );
+      */
```

Comment:
Do you know what's the best way to test that `InputIt` is "and iterator type"?

https://en.cppreference.com/w/cpp/container/vector/vector

describes the exact problem that I ran into with this. That without any type requirements, this constructor is preferred over the one defined just before it.

Unfortunately STL implements its type requirements with non-public code. :frowning: So for now I rather just removed these lines until I get a better idea of how to put in a bit of SFINAE on this function...

Reply:
I'm not sure what you mean. If you leave the second constructor uncommented it will implicitly upcast and use that one instead of the one on line 75? In that case, you should make the constructor of InputIt explicit.